src/main/cpp/mechanisms/noteManager/decoratormods/noteManager.cpp

```diff
@@ -233,19 +236,33 @@ void noteManager::Update(RobotStateChanges::StateChange change, int value)
 
 double noteManager::GetRequiredLaunchAngle()
 {
+	frc::DriverStation::Alliance allianceColor = FMSData::GetInstance()->GetAllianceColor();
 	double distanceFromTarget = 3.5;
 	double launchAngle = 0;
+	frc::Pose3d fieldElementPose = frc::Pose3d{};
+	auto config = ChassisConfigMgr::GetInstance()->GetCurrentConfig();
+	auto chassis = config != nullptr ? config->GetSwerveChassis() : nullptr;
+	frc::Pose2d chassisPos = frc::Pose2d();
 
 	if (HasVisionTarget())
 	{
 		distanceFromTarget = GetVisionDistance().to<double>();
+	}
+	else if (chassis != nullptr)
+	{
+		fieldElementPose = allianceColor == frc::DriverStation::Alliance::kRed ? frc::Pose3d{FieldConstants::GetInstance()->GetFieldElement(FieldConstants::FIELD_ELEMENT::RED_SPEAKER)} /*load red speaker*/ : frc::Pose3d{FieldConstants::GetInstance()->GetFieldElement(FieldConstants::FIELD_ELEMENT::BLUE_SPEAKER)}; /*load blue speaker*/
+		chassisPos = chassis->GetPose();
 
-		launchAngle = 80.0 + (-44.2 * distanceFromTarget) + (6.09 * distanceFromTarget * distanceFromTarget);
+		distanceFromTarget = sqrt(pow((fieldElementPose.X() - chassisPos.X()).to<double>(), 2) + pow((fieldElementPose.Y() - chassisPos.Y()).to<double>(), 2));
 	}
+
+	launchAngle = 80.0 + (-44.2 * distanceFromTarget) + (6.09 * distanceFromTarget * distanceFromTarget);
```

Comment:
Can we replace the magic numbers with constants to make this easier to read?

Reply:
Leaving until we make a new function